Proposal: Create assert_response_schema test helper

maurogeorge opened the pull request:
It is a common pattern to use JSON Schema to validate a API response [1], [2]
and [3].

This patch creates the `assert_response_schema` test helper that helps people do
this kind of validation easily on the controller tests.

This makes sense to you guys? If make sense I can continue with CHANGELOG, Docs and so on.

[1]: https://robots.thoughtbot.com/validating-json-schemas-with-an-rspec-matcher
[2]: https://github.com/sharethrough/json-schema-rspec
[3]: https://github.com/rails-api/active_model_serializers/issues/1011#issuecomment-127608121

bf4 commented:
Wow, this is awesome! I think I made a issue to add this.. specifically for json-api, but in general is also great!

I have a few comments on implementation, but am generally wow!

maurogeorge commented:
Hi @bf4 great to hear that :smiley: 
I added some docs, now waiting for your review.

bf4 commented on lib/active_model/serializer/assertions.rb:
```diff
@@ -0,0 +1,16 @@
+module ActiveModel
+  class Serializer
+    module Assertions
+      def assert_response_schema(schema_path = nil)
```
should be under a test namespace if it's for tests, e.g. `ActiveModelSerializers::Test::Schema`

And would be good for the method not to be dependent on where it's included 

e.g.

```ruby
module Schema
  module_function

  def validate!(schema_full_path, json, strict = true)
    JSON::Validator.validate!(schema_full_path, json, strict: strict) # is this keyword args? we can't support that yet
  end

  def validate_response_schema!(response, schema_path = nil)
    validate!(schema_full_path_for(response.body, schema_path))
  end

  def schema_full_path_for(response, schema_path = nil)
        controller_path = response.request.filtered_parameters[:controller]
        action = response.request.filtered_parameters[:action]
        schema_directory = ActiveModel::Serializer.config.schema_path
        schema_path ||= "#{controller_path}/#{action}.json"
        "#{schema_directory}/#{schema_path}"
  end

```

bf4 commented on lib/active_model_serializers.rb:
```diff
@@ -51,6 +52,7 @@ def silence_warnings
   ActionDispatch::Reloader.to_prepare do
     ActiveModel::Serializer.serializers_cache.clear
   end
+  ActionController::TestCase.send(:include, ActiveModel::Serializer::Assertions)
```
This should be in a Railtie or something like that..

bf4 commented:
Sorry for delay in review this.  There'll be a good rebase ...

bf4 commented:
ref: https://github.com/rails-api/active_model_serializers/issues/1162

maurogeorge commented:
Thanks for the review, I think it is better hold the working on this until the https://github.com/rails-api/active_model_serializers/pull/1291 be merged, since both define instrumentation and the last one is more mature.

First I will work on the #1291 until the merge and later I back to this one.

bf4 commented:
wanna take another look at this? I like this PR

maurogeorge commented:
I think it is better hold the working on this until the https://github.com/rails-api/active_model_serializers/pull/1322 be merged, since both define instrumentation and the last one is more mature.

bf4 commented:
instrumentation PR is merged!

malandrina commented:
This is a cool idea!

I wanted to chime in and share some thoughts about compatibility with other JSON Schema-based tools. Something we discovered recently with [`json_matchers`](https://github.com/thoughtbot/json_matchers) at thoughtbot is that `json_matchers` currently isn’t compatible with Heroku’s JSON Schema-based tools, e.g. [`prmd`](https://github.com/interagent/prmd) and [`committee`](https://github.com/interagent/committee), because it makes different assumptions about the organization and structure of the schemata.

It’s likely you’ll run into the same issue with the implementation you have here. If you’re interested in ensuring compatibility with Heroku’s tooling you should consider using [`json_schema`](https://github.com/brandur/json_schema) to parse and validate schemata and organize the schemata according to Heroku’s [conventions](https://github.com/interagent/prmd/blob/master/docs/schemata.md).

maurogeorge commented:
@malandrina thanks a lot for you thoughts here :heart: 

I missed this on README of [json-schema](https://github.com/ruby-json-schema/json-schema) is said that this does not handle any JSON Hyper-Schema and prmd uses JSON Hyper-Schema.
The [json_schema](https://github.com/brandur/json_schema) you recommended handle this. So yes I will change this to use json_schema.

When you said it makes different assumptions about the organization and structure of the schemata,  [the recommended way to organize the individual resource schema](https://github.com/interagent/prmd#file-layout) is under `docs/schema/schemata`. I created a option to define the [path here](https://github.com/rails-api/active_model_serializers/pull/1270/files#diff-3a2aebf9cd60f74f3a64262a18a3962fR70) so if we change this to `ActiveModel::Serializer.config.schema_path = 'docs/schema/schemata'` we can use the same json files to the prmd and comittee right?

Thanks

malandrina commented:
@maurogeorge no problem!

Based on my experience with [this `json_matchers` issue](https://github.com/thoughtbot/json_matchers/issues/25) (apologies for not having referenced it directly in my previous comment), I'm not confident that making the root path for the schemata configurable is enough to achieve compatibility with Heroku's tools. I believe users of ActiveModelSerializers will have to structure their schemata according to Heroku's conventions (e.g. [including required metadata](https://github.com/interagent/prmd/blob/master/docs/schemata.md#meta-data) and [links](https://github.com/interagent/prmd/blob/master/docs/schemata.md#links))  in addition to storing the schema files in `docs/schema/schemata`.

maurogeorge commented:
@malandrina I updated to use the `json_schema`. To test the heroku integration  I create a hyper-schema using the prmd and test against it. So now it is possible to do some simple validation using a simple json or using the hyper-schema format.

With this I think the users can structure their schemata according to Heroku's conventions and storing the schema files in `docs/schema/schemata`. You think this is ok or I miss something.

Thanks

malandrina commented:
@maurogeorge your process seems solid. :+1: The only other I thing I would suggest thinking about is use of the `$ref` keyword and [JSON Pointers](http://spacetelescope.github.io/understanding-json-schema/UnderstandingJSONSchema.pdf) in the simple (i.e. non-hyper) schemata.

If you would like for users of the matcher to be able to use JSON Pointers in non-hyper schemata, I'd suggesting writing a test or two to validate your assumptions about how JSON Pointers should be formatted and how `json_schema` dereferences them. This could be as simple as adding a couple `$ref`s to `test/support/schemas/custom/show.json`.

If you don't expect users to include JSON Pointers in non-hyper schemata, I'd recommend including that information in the documentation.

bf4 commented:
I noticed that the `hana` gem includes a json pointer implementation (which is pretty simple) https://github.com/tenderlove/hana/blob/5da3b986cd529c680c96aacf20fc7f3c223d8102/lib/hana.rb#L4-L37 compared to https://github.com/tent/json-pointer-ruby/blob/ade5fea30acd9c85c3886db62e83d68e3b5919bb/lib/json-pointer.rb#L3-L220

bf4 commented on docs/howto/test.md:
```diff
@@ -0,0 +1,78 @@
+# How to test
+
+## Test helpers
```
This is an end-user's guide to testing, right? (vs. in CONTRIBUTING.md how to develop on AMS).

In any case, this is :100:

maurogeorge replied:
@bf4 thanks! Yes, this is the idea.

bf4 commented on lib/active_model/serializer/railtie.rb:
```diff
@@ -19,5 +19,7 @@ class Railtie < Rails::Railtie
       require 'generators/serializer/resource_override'
     end
+
+    ActionController::TestCase.send(:include, ActiveModelSerializers::Test::Schema)
```
nice

bf4 replied:
`if Rails.env.test?` ?

maurogeorge replied:
I don't think we need this `if Rails.env.test?` Active Job and Action Mailer for example does not do that as you can see [here](https://github.com/rails/rails/blob/d7add5352e7c8b31850431bb93acb56d93c5dc64/activejob/lib/active_job.rb#L34), [here](https://github.com/rails/rails/blob/d7add5352e7c8b31850431bb93acb56d93c5dc64/activejob/lib/active_job/test_case.rb), [here](https://github.com/rails/rails/blob/d7add5352e7c8b31850431bb93acb56d93c5dc64/actionmailer/lib/action_mailer.rb#L46-L47) and [here](https://github.com/rails/rails/blob/d7add5352e7c8b31850431bb93acb56d93c5dc64/actionmailer/lib/action_mailer/test_helper.rb).

bf4 replied:
tl;dr let's verify if we can load AMS in production mode without the conditional

There's an issue about this in the ams issues.. I can go find it, but if you eager load the test **framework** in non-test, it will blow up trying to require minitest among other things.  

The difference here and the example links (thanks! I love how you research this stuff), is that e.g. we're patching an existing testcase, not writing our own. 

e.g. see [`class ActiveJobTestCaseTest < ActiveJob::TestCase`](https://github.com/rails/rails/blob/7f18ea14c893cb5c9f04d4fda9661126758332b5/activejob/test/cases/test_case_test.rb)

[ActiveJob::TestHelper](https://github.com/rails/rails/blob/d7add5352e7c8b31850431bb93acb56d93c5dc64/activejob/lib/active_job/test_helper.rb) has no explicit dependency on a test framework being loaded.  

[`ActiveSupport::TestCase`](https://github.com/rails/rails/blob/d7add5352e7c8b31850431bb93acb56d93c5dc64/activejob/lib/active_job/test_case.rb) does inherit from `ActiveSupport::TestCase`, but it is only ever loaded in test when subclassing it.

I think I verified this in console. Here's one of the issues I was thinking of https://github.com/rails-api/active_model_serializers/issues/814

bf4 commented on test/support/schemas/hyper_schema.json:
```diff
@@ -0,0 +1,93 @@
+{
+  "$schema": "http://json-schema.org/draft-04/hyper-schema",
```
are you using something to generate these?

might also be nice to add  some JSON API stuff even like in https://github.com/rails-api/active_model_serializers/pull/1301#discussion_r44242510 (which should probably be merged, I guess)

maurogeorge replied:
@bf4 yes, this was created using the prmd gem. I will take a look at your comment.

maurogeorge commented:
@malandrina thanks for the tip, I updated the code https://github.com/maurogeorge/active_model_serializers/commit/0852b49448b222a3e6cf2b3e35fe3a3d545fb36e do you think this is enough?

Thanks

bf4 commented:
@maurogeorge Looks like CI wants you to rebase off of master and force push

This PR is so awesome.  I love looking at it!

maurogeorge commented:
@bf4 thanks for the tip, I will rebase and push again.

I added the [JSON API as you mentioned](https://github.com/rails-api/active_model_serializers/pull/1270#discussion_r46382912) https://github.com/maurogeorge/active_model_serializers/commit/0f6a3461fc1067be1acf55208e38c56305222055 

:heart:

maurogeorge commented:
@bf4 could you please review when you have some time? The build will be [broken at Ruby 1.9, but this is related with the json_schema](https://travis-ci.org/rails-api/active_model_serializers/jobs/94753014#L181) that [uses keyword arguments](https://github.com/brandur/json_schema/blob/61d60d27fdc8c75ef5b1686a4744c7a31a3a6fb7/lib/json_schema/schema_error.rb#L23). I will try to send a patch the next time I work on this.

bf4 commented on active_model_serializers.gemspec:
```diff
@@ -46,6 +46,8 @@ Gem::Specification.new do |spec|
     # activesuport
     # activemodel
 
+  spec.add_runtime_dependency 'json_schema'
```
- [ ] Breaks Ruby 1.9, JRuby 1.9-mode due to using keyword arguments

bf4 replied:
We've removed 1.9 support, so it's all good

bf4 commented on docs/howto/test.md:
````diff
@@ -0,0 +1,78 @@
+default schema path in a initializer.
+
+```ruby
+ActiveModel::Serializer.config.schema_path = `spec/support/schemas`
````
I think we should prefer `ActiveModelSerializers.config.schema_path = 'spec/support/schemas'` at this point

Also 

```plain
schema_path = `spec/support/schemas`
should be
schema_path = 'spec/support/schemas'
```

bf4 commented on lib/active_model/serializer/configuration.rb:
```diff
@@ -20,6 +20,7 @@ def config.array_serializer
 
         config.adapter = :attributes
         config.jsonapi_resource_type = :plural
+        base.config.schema_path = 'test/support/schemas'
```
`s/base.config/config`

bf4 commented on lib/active_model_serializers/test/schema.rb:
```diff
@@ -0,0 +1,62 @@
+        end
+
+        def schema_directory
+          ActiveModel::Serializer.config.schema_path
```
`s//ActiveModelSerializers`

bf4 commented on lib/active_model_serializers/test/schema.rb:
```diff
@@ -0,0 +1,62 @@
+        end
+
+        def schema_data
+          JSON.parse(File.read(schema_full_path))
```
might be nice to allow other JSON parsers via e.g. 

```ruby
def load_json(json_string)
  JSON.load(json_string) # or JSON.parse.  I just like the load/dump marshal semantics
end
```

Also, we'll want to check for or otherwise handle the non-existence, readability, parseability of the schema file.  

`File.exist?` `File.readable?` `rescue JSON::ParserError`

maurogeorge replied:
@bf4 sorry I didn't understand the `load_json` idea, could you please show me a example of the public API? Today we only expose this via the `assert_response_schema`. And what is the use case to this, the `assert_response_schema` purpose is to be used on controllers tests so I don't think we need to add the ability to load custom data.

Thanks for the review

bf4 replied:
pseudo-code (not executed), sorry for delay

```ruby
def schema_path
  load_json_file(schema_full_path)
end

def response_body
  load_json(response.body)
end

def add_schema_to_document_store
  Dir.glob("#{schema_directory}/**/*.json").each do |path|
    schema_data = JSON.parse(File.read(path))
    extra_schema = JsonSchema.parse!(schema_data)
    document_store.add_schema(extra_schema)
  end
end

def load_json(json_string)
  JSON.load(json_string) # or JSON.parse.  I just like the load/dump marshal semantics
rescue JSON::ParserError => e
  raise JSON::ParserError, "Better error message", e.backtrace
end

def load_json_file(path)
  if File.exist?(path) && File.readable?(path)
    load_json(File.binread(path))
  else
    raise Errno::NotFound, "json file does not exist at #{path}"
  end
end
```

bf4 commented:
I'd like to look through it more carefully, but still looks great :)

malandrina commented:
@maurogeorge I had one more thought -- https://github.com/maurogeorge/active_model_serializers/commit/0852b49448b222a3e6cf2b3e35fe3a3d545fb36e#commitcomment-14792738 -- but otherwise you've addressed all the concerns I raised.

Looking good!

maurogeorge commented:
@malandrina thanks for the review.
I answered your comments here https://github.com/maurogeorge/active_model_serializers/commit/0852b49448b222a3e6cf2b3e35fe3a3d545fb36e#commitcomment-14885685

maurogeorge commented:
@bf4 I updated the code, when you have some time please take a look.

Thanks

bf4 commented:
Will do.  Probably tomorrow 

B mobile phone

> On Dec 10, 2015, at 3:50 PM, Mauro George <notifications@github.com> wrote:
> 
> @bf4 I updated the code, when you have some time please take a look.
> 
> Thanks
> 
> —
> Reply to this email directly or view it on GitHub.
>

bf4 commented:
@maurogeorge I want to help get this across the finish line. Do you have time we can chat, just to minimize any back and forth.  I think only important thing is really to rebase against master (and apply the docs in the right spot, now that their de-duped)

maurogeorge commented:
@bf4 sorry I really don't know when I have some time to do that, makes sense to you to merge when you believe that are ok and do the changes in a new PR? Or we can continue asynchronously.

I rebased againtst the master. Could you please point to me where is the right place to add the docs?

I don't think we can merge this right know, first we need to fix the pointer, It seems that this is not working properly, more info here https://github.com/maurogeorge/active_model_serializers/commit/0852b49448b222a3e6cf2b3e35fe3a3d545fb36e#commitcomment-14792738 and https://github.com/brandur/json_schema/issues/41#issuecomment-164536458 I will take a look at this right now.

bf4 commented:
@maurogeorge 

> I don't think we can merge this right know, first we need to fix the pointer, It seems that this is not working properly, more info here maurogeorge@0852b49#commitcomment-14792738 and brandur/json_schema#41 (comment) I will take a look at this right now.

Ok, I'll be in chat if you'd like

maurogeorge commented:
Hi @malandrina I think I made [your last recommendation](https://github.com/maurogeorge/active_model_serializers/commit/0852b49#commitcomment-14792738) on this commit https://github.com/rails-api/active_model_serializers/commit/f1e6b6a174fc937f88e65dc384a04309e302baa7 now we have the JSON Pointers working properly.

@bf4 could you please review? 

Thanks a lot guys :heart:

bf4 commented:
Looks like some failures are legit.  I'm just going to merge the fix to master and make our lives a little saner

```diff
FAIL ActiveModelSerializers::Test::SchemaTest#test_with_a_non_existent_file (0.01s)
  --- expected
  +++ actual
  @@ -1 +1 @@
  -"No such file or directory @ rb_sysopen - test/support/schemas/non-existent.json"
  +"No such file or directory - test/support/schemas/non-existent.json"
  /home/travis/build/rails-api/active_model_serializers/test/active_model_serializers/test/schema_test.rb:108:in `test_with_a_non_existent_file'
```

well, `@ rb_sysopen` so we just want to make the checking less strict

bf4 commented:
I just merged https://github.com/rails-api/active_model_serializers/pull/1384 since the failures on master were driving me nuts (and probably everyone else rebasing off of master these days)

bf4 commented on active_model_serializers.gemspec:
```diff
@@ -46,6 +46,8 @@ Gem::Specification.new do |spec|
     # activesupport
     # activemodel
 
+  spec.add_runtime_dependency 'json_schema'
```
It's kind of interesting that this is a runtime dependency intended for usage in client tests.

bf4 commented on lib/active_model_serializers/test/schema.rb:
```diff
@@ -0,0 +1,77 @@
+module ActiveModelSerializers
+  module Test
+    module Schema
+      def assert_response_schema(schema_path = nil, message = nil)
```
would you mind adding comments to the effect of what nil schema_path/message mean and what types of data they are? Yardoc or pseudo-yardoc is preferred.

```ruby
# A Minitest Assertion that the response is valid per the schema
# which uses the given error message
# @params schema_path [String] blah nil blah else
# @params message [String] blah nil blah else
# @example
#   get :something
#   assert_response_schema
```

and the matcher just returns bool?

bf4 commented on lib/active_model_serializers/test/schema.rb:
```diff
@@ -0,0 +1,77 @@
+
+        ActiveModelSerializers.silence_warnings do
+          attr_reader :document_store
+        end
```
this silence isn't necessary if the reader is in the protected scope, so that's what I've been doing lately. (it's a ruby bug)

bf4 commented on test/active_model_serializers/test/schema_test.rb:
```diff
@@ -0,0 +1,112 @@
+      end
+
+      def test_with_a_non_existent_file
+        message = 'No such file or directory @ rb_sysopen - test/support/schemas/non-existent.json'
```
error message is too specific. some systems don't have the `rb_sysopen` string.  Better to match /No such file/ and the file path

bf4 commented on docs/howto/test.md:
```diff
@@ -0,0 +1,78 @@
+
+## Test helpers
+
+AMS provides a `assert_response_schema` method to be used on your controller tests to
```
s/AMS/ActiveModelSerializers

bf4 commented on docs/howto/test.md:
```diff
@@ -0,0 +1,78 @@
+
+AMS provides a `assert_response_schema` method to be used on your controller tests to
+assert the response against a [JSON Schema](http://json-schema.org/). Let's take
+a look in a example.
```
Probably best to note it is minitest only to head off issues

bf4 commented on lib/active_model/serializer/railtie.rb:
```diff
@@ -19,5 +19,9 @@ class Railtie < Rails::Railtie
     end
+
+    if Rails.env.test?
+      ActionController::TestCase.send(:include, ActiveModelSerializers::Test::Schema)
```
Too bad include is still private in Ruby 2.0 :(

bf4 commented:
@maurogeorge Looks great.  I left a few last comments, some are actionable, some are just follow up discussion, and it's good to merge.

maurogeorge commented:
@bf4 I updated the code, please review when you have some time.

Thanks

bf4 commented on docs/howto/test.md:
````diff
@@ -0,0 +1,126 @@
+```
+
+```json
+# show.json
````
invalid json :trollface:

maurogeorge replied:
:smile:

bf4 commented on lib/active_model_serializers/test.rb:
```diff
@@ -0,0 +1,5 @@
+module ActiveModelSerializers
+  module Test
+    autoload :Schema, 'active_model_serializers/test/schema'
```
- [ ] elsewhere we're using `  extend ActiveSupport::Autoload` in place of the ruby autoload.  we can do this in a followup PR

bf4 commented on lib/active_model_serializers/test/schema.rb:
```diff
@@ -0,0 +1,98 @@
+      end
+
+      MissingSchema = Class.new(Errno::ENOENT)
+      InvalidSchemaError = Class.new(StandardError)
```
- [ ] these usually go at the top of the file.

bf4 replied:
Might want to subclass `s/StandardError/JSON::ParserError`

bf4 commented on test/active_model_serializers/test/schema_test.rb:
```diff
@@ -0,0 +1,128 @@
+      end
+
+      def test_with_a_non_existent_file
+        message = %r{.* - No Schema file at test/support/schemas/non-existent.json}
```
- [ ] `.*` at start of regex probably unnecessary, but it's in a test and doesn't have much impact

maurogeorge replied:
I used the `.*` because I got the following outputs from different ruby versions:

- Ruby 2.2.3:
  - No such file or directory - No Schema file at test/support/schemas/non-existent.json

- rbx-2
  - unknown error - No Schema file at test/support/schemas/non-existent.json

It can be updated to something like `%r{(No such file or directory|unknown error) - No Schema file at test/support/schemas/non-existent.json}`. But I think we can read the actual like: something generated by the Ruby and our good error message.

bf4 commented:
Also, related issue: https://github.com/rails-api/active_model_serializers/issues/1162

maurogeorge commented:
@bf4 Thanks for the review.

I updated the code following your comments. We are back to green after drop Ruby 1.9 on Travis.

The AppVeyor is broken because we are running Ruby 1.9.3 on it, I can't found a dot file to change this, I imagine this can be managed via some interface, so we need to change this.

Thanks

maurogeorge commented:
@bf4 I sent a PR to fix AppVeyor the https://github.com/rails-api/active_model_serializers/pull/1395 I found for a dot file, but it is not a dot file is `appveyor.yml` only :smile:

maurogeorge commented:
@bf4 I updated the code [following your last comments](https://github.com/rails-api/active_model_serializers/pull/1270#issuecomment-166991852) and rebased against the master.

We are good to :shipit: ?

Thanks

bf4 commented:
Looks good but I'm tired right now and want to review more carefully. Did I miss why json_schema is a runtime dependency instead of development?

@malandrina thoughts?

cc @rails-api/ams this is a good one to look at

maurogeorge commented:
@bf4 thanks, no problem take a rest :smiley: 

I think if we keep the `json_schema` as a development dependency we will need to ask for people to install this, before use the test helper.

bf4 commented:
Oh right.  We could warn inside a method like rails attr_encrypted so rails doesn't need to specify bcrypt 

B mobile phone

> On Jan 6, 2016, at 3:09 PM, Mauro George <notifications@github.com> wrote:
> 
> @bf4 thanks, no problem take a rest 
> 
> I think if we keep the json_schema as a development dependency we will need to ask for people to install this, before use the test helper.
> 
> —
> Reply to this email directly or view it on GitHub.
>

malandrina commented on docs/howto/test.md:
```diff
@@ -0,0 +1,126 @@
+
+ActiveModelSerializers provides a `assert_response_schema` method to be used on your controller tests to
+assert the response against a [JSON Schema](http://json-schema.org/). Let's take
+a look in a example.
```
Grammar: "a example" => "an example"

malandrina commented on docs/howto/test.md:
````diff
@@ -0,0 +1,126 @@
+```
+
+To test the `posts#show` response of this controller we need to create a file in
+`test/support/schemas/posts/show.json` the helper uses a convention to the name
````
Looks like this should be two sentences, i.e.

> To test the `posts#show` response of this controller we need to create a file named `test/support/schemas/posts/show.json`. The helper uses a naming convention to locate the file.

malandrina commented on docs/howto/test.md:
```diff
@@ -0,0 +1,126 @@
+
+### Load a custom schema
+
+If we need to use other schema, for example when we have a namespaced API that
```
"other" => "another"

malandrina commented on docs/howto/test.md:
````diff
@@ -0,0 +1,126 @@
+```
+### Change the schema path
+
+By default all schemas are created at `test/support/schemas` if we are using
````
Grammar: There should be a period after `test/support/schemas

malandrina commented on docs/howto/test.md:
```diff
@@ -0,0 +1,126 @@
+
+By default all schemas are created at `test/support/schemas` if we are using
+RSpec for example we can change this to `spec/support/schemas` defining the
+default schema path in a initializer.
```
Grammar: "a initializer" => "an initializer"

malandrina commented:
I left some comments with suggested improvements to the documentation but overall this seems reasonable to me. It's great that this will be compatible with Heroku's tooling. :+1:

maurogeorge commented:
@bf4 Liked the idea, I update the code following the same idea [Rails does](https://github.com/rails/rails/blob/869a90512f36b04914d73cbf58317d953caea7c5/activemodel/lib/active_model/secure_password.rb#L57-L62).

@malandrina thanks for the review, I updated the docs accordingly.

bf4 commented:
@maurogeorge rebase and good to merge, I think.  Wanna join the commiters team?

maurogeorge commented:
@bf4 rebased, :shipit: 

:blush: Yeah, I like the idea to join the commiters team. Probably I will leave the open source world for a while, but later I can continue working on the project.

bf4 commented:
@maurogeorge awesome. I'll have @joaomdmoura add you. (I thought I could, but I guess not.)

joaomdmoura commented:
@maurogeorge You are invited! :heart: would be nice to have you on slack so that I can add you on the maintainer channel :smile:

maurogeorge commented:
@bf4 @joaomdmoura thanks :heart: 

João as I said in the https://github.com/rails-api/active_model_serializers/pull/1270#issuecomment-171427499, I will leave the open source world for a while, but later I can continue working on the project and help more the project.

@bf4 :shipit: ?

bf4 commented:
Merged by ac130537908904f2a2f26ff0cf8b0848fe4fd196

maurogeorge commented:
Thanks for the merge :+1:
